dpsim/include/dpsim/Interface.h

```diff
@@ -34,38 +34,78 @@ namespace DPsim {
-		virtual CPS::Attribute<Real>::Ptr importReal(UInt idx) = 0;
-		virtual CPS::Attribute<Bool>::Ptr importBool(UInt idx) = 0;
-		virtual CPS::Attribute<Complex>::Ptr importComplex(UInt idx) = 0;
-		virtual CPS::Attribute<Complex>::Ptr importComplexMagPhase(UInt idx) = 0;
-
-		virtual void exportAttribute(CPS::AttributeBase::Ptr attr, Int idx);
-		virtual void exportInt(CPS::Attribute<Int>::Ptr attr, UInt idx, const std::string &name="", const std::string &unit="") = 0;
-		virtual void exportReal(CPS::Attribute<Real>::Ptr attr, UInt idx, const std::string &name="", const std::string &unit="") = 0;
-		virtual void exportBool(CPS::Attribute<Bool>::Ptr attr, UInt idx, const std::string &name="", const std::string &unit="") = 0;
-		virtual void exportComplex(CPS::Attribute<Complex>::Ptr attr, UInt idx, const std::string &name="", const std::string &unit="") = 0;
-
-		/** Read data for a timestep from the interface and passes the values
-		 * to all registered current / voltage sources.
-		 */
-		virtual void readValues(bool blocking = true);
-
-		/** Write all exported values to the interface. Called after every timestep.
-		 * @param model Reference to the system model which should be used to
-		 * calculate needed voltages.
-		 */
-		virtual void writeValues() = 0;
-
-		virtual CPS::Task::List getTasks() = 0;
+		virtual void importAttribute(CPS::AttributeBase::Ptr attr);
+		virtual void exportAttribute(CPS::AttributeBase::Ptr attr);
+
+		//Function used in the interface's simulation task to read all imported attributes from the queue
+		//Called once before every simulation timestep
+		virtual void readValuesFromQueue();
+		//Function used in the interface's simulation task to write all exported attributes to the queue
+		//Called once after every simulation timestep
+		virtual void writeValuesToQueue();
+
+		//Function that will be called on loop in its separate thread.
+		//Should be used to read values from the environment and push them into the queue
+		virtual void readValuesFromEnv() = 0;
+
+		//Function that will be called on loop in its separate thread.
+		//Should be used to read values from the queue and write them to the environment
+		virtual void writeValuesToEnv() = 0;
+
+		virtual CPS::Task::List getTasks();
 
 		bool shouldSyncOnSimulationStart() const {
 			return mSyncOnSimulationStart;
 		}
 
+		virtual ~Interface() {
+			if (mOpened)
+				close();
+		}
+
 	protected:
-		bool mSyncOnSimulationStart = false;
+		CPS::AttributeBase::List mExportAttrs, mImportAttrs;
+		CPS::Logger::Log mLog;
+		bool mBlockOnRead;
+		bool mSyncOnSimulationStart;
+		UInt mDownsampling;
+		String mName;
+		bool mOpened;
+
+	public:
+		class PreStep : public CPS::Task {
+		public:
+			PreStep(Interface& intf) :
+				Task(intf.mName + ".Read"), mIntf(intf) {
+				for (auto attr : intf.mImportAttrs) {
+					mModifiedAttributes.push_back(attr);
+				}
+				//TODO: Is this necessary / what effect does a dependency on external have?
```

Comment:
As mentioned before, the external dependency avoids that the scheduler drops the task.